software/spmd/pc_profiler/profiler.S

```diff
@@ -0,0 +1,232 @@
+
+  Registers used:
+  The program being profiled is free to use any registers within if there are no other imposed constraints.
+  The profiler uses the following registers - 
+  t0 --> Base address of the histogram array (actual value stored in dmem[1])
+  t1 --> Scratchpad (actual value stored in dmem[0])
+*/
+
+#include "bsg_manycore_arch.h"
+#include "bsg_manycore_asm.h"
+
+// Useful to control the amount of space required to simulate the use of the profiler.
+// We want 16MB when running on actual hardware.
+#ifndef HIST_SPACE
+#define HIST_SPACE 16777216
+#endif
+
+#ifndef PROFILE
+#define PROFILE 0
+#endif
+
+#if PROFILE > 0
+/* 
+  Data section:
+  Defines a 2 element (8 byte) array to store the base pointer of the histogram array and for use by the interrupt handler to spill some registers
+  dmem[0] = histogram base pointer, temporary storage
+  dmem[1] = temporary storage
+*/
+.section .dmem.interrupt, "aw"
+  _interrupt_arr: .space 8, 0x0
+
+/*
+  DRAM section:
+  Defines a 16MB (array size is configurable) chunk in DRAM to store the histogram.
+
+  Lingering question(s):
+  1. Assigned space per PC (i.e bin capacity) as 32-bits. Is this enough, too much or too less?
+  2. Base pointer in DRAM (for now it is mapped to start of DRAM) [Consider static assignment in linker]
+*/
+.section .dram, "aw"
+.globl _histogram_arr
+  _histogram_arr: .space HIST_SPACE, 0x0
+#endif
+
+/*
+  Startup:
+  This is very similar to the code in software/spmd/common/crt.S and adds only the interrupt service routine (ISR) and sets up the DMEM with the histogram base address. 
+*/
+.section .crtbegin
+.globl _start
+
+#if PROFILE > 0
+/*
+  Remote interrupt handler (should map to address 0)
+*/
+_remote_interrupt:
+  j _remote_interrupt_handler
+
+/*
+  Trace interrupt handler (should map to address 4)
+  Clears the pending trace interrupt bit in the MIP CSR and atomically increments the histogram bin count based on the MEPC CSR value
+
+  The bin index in the histogram can be found using the following equation index = base ptr + EPC
+  (Although EPC is 4-byte aligned, we are assigning 4 bytes per bin, so we don't need to do EPC >> 2)
+
+  Lingering concerns:
+  Since we are using the MEPC, we don't have the PC of the first instruction in the program. 
+  (Probably does not matter because it's just 1 datapoint)
+*/
+_trace_interrupt:
+  // Save a register to dmem[1]
+  sw t0, 4(zero)
+
+  // clear mip.trace
+  li t0, 0x20000
+  csrrc zero, mip, t0
+
+  // load address of histogram array
+  lw t0, 0(zero)
+  // Store another register to dmem[0]
+  sw t1, 0(zero)
+
+  // read the mepc register
+  csrr t1, mepc
+  // find the correct index in the histogram array
+  add t0, t0, t1
+
+  // atomically increment the value at the computed index
+  li t1, 1
+  // We don't care about the result hence assigned to zero register
+  amoadd.w.aq zero, t1, (t0)
+
+  // remap t0 to histogram array base address
```

Comment:
rather than rematerializing the base address, so you can store it back, I suggest that you instead just use one more word of data memory for your second temporary. you reduce icache footprint by 3 instructions, and increase tracing performance by 3/(15+stall cycles (6?)).

Reply:
Isn't that going away from the spec that we could only use the first 2 entries in the dmem for interrupts? If it's not a strict requirement then that's great!

Just out of curiosity, how did you compute the divisor for the expression you mentioned?

Reply:
I know there’s a slippery slope argument, but maybe just go to four right now so that when we add a fast remote interrupt handler we have a little more breathing room?

Reply:
I am curious why we are hardcoding that address for interrupts? Does the hardware treat those locations differently? Is it just we can avoid using the extra instructions to load relocatables?

Reply:
Right now, the hardware expects interrupt handlers at these addresses as per https://github.com/bespoke-silicon-group/bsg_manycore/blob/43ba88eadbaebc7a446fcb60077dece3aaea3f66/v/vanilla_bean/bsg_vanilla_pkg.v#L323

Maybe @tommy could throw some light why these need to be hardcoded?

Reply:
So there seems to be two things that are hardcoded:
1. The PC address that is set when an interrupt is triggered
2. The private data in DMEM that is reserved for the interrupt handler's use

I understand why (1) - I'm a little less clear on (2).

Reply:
interrupt jump address can be hardcoded, because it's good enough to do what we want, and area-efficient. For reserved space in DMEM for interrupt handler, there is nothing in hardware that hardcodes that. But we put it there, so that programmer can access it using x0 as base address (e.g. sw x3, 0(x0) or sw x3, 4(x0))

Reply:
ok - so (2) is for efficient addressing of the interrupt data.

Reply:
Yeah normally this is done using a swap with a dedicated scratch CSR and interrupt stack in memory.

 This approach is a scratch register and 1-2 address stack (located in DMEM)